Allow copying manifold ids in GridGenerator::merge_triangulations

masterleinad opened the pull request:
This PR allows copying all `manifold_ids` when calling `GridGenerator::merge_triangulations`. This is quite helpful when creating `Triangulation` objects from other ones that have a manifold attached by default.

davydden commented on source/grid/grid_generator.cc:
```diff
@@ -3818,6 +3821,7 @@ namespace GridGenerator
         for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell; ++v)
           this_cell.vertices[v] = cell->vertex_index(v);
         this_cell.material_id = cell->material_id();
+        this_cell.manifold_id = cell->manifold_id();
```
while there, could you please switch to range loop above?

davydden commented on source/grid/grid_generator.cc:
```diff
@@ -3833,12 +3837,128 @@ namespace GridGenerator
+                  subcell_data.boundary_lines.push_back(
+                    std::move(boundary_line));
+                }
+            for (typename Triangulation<dim, spacedim>::line_iterator line =
```
while not too much saving, one could add a single lambda function which takes `tria` and `offset` and call it twice on `tria_1` (offset=0) and `tria_2` (offset=n_vertces()).

davydden commented on source/grid/grid_generator.cc:
```diff
@@ -3833,12 +3837,128 @@ namespace GridGenerator
+                        std::move(boundary_line));
+                    }
+              }
+            for (typename Triangulation<dim, spacedim>::face_iterator face =
```
one can make a lambda here as well.

davydden commented on tests/grid/merge_triangulations_05.output:
```diff
@@ -0,0 +1 @@
+
```
you probably forgot to add output.

masterleinad replied:
Never forget to flush `deallog`! Fixed now.

masterleinad commented:
@davydden I addressed your comments. Using lambdas is nicer indeed.

> It looks like the code handles quads and lines with different boundary ids but its not clear to me what happens when the same face (on different triangulations) has a different manifold id. Is this handled somewhere?

Yes, now it is. Have a look at https://github.com/dealii/dealii/pull/6898/files#diff-3fc86b38720e340f14953e51dfe4ea5aR3123 and https://github.com/dealii/dealii/pull/6898/files#diff-3fc86b38720e340f14953e51dfe4ea5aR3275. We ensure that the manifold ids are consistent.

bangerth commented on doc/news/changes/minor/20180708DanielArndt:
```diff
@@ -0,0 +1,3 @@
+New: GridGenerator_merge_triangulations gained an option to copy manifold ids.
```
Should be `GridGenerator::...`

bangerth commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -984,11 +984,11 @@ namespace GridGenerator
-   * will currently have to set boundary indicators again by hand in the
-   * output triangulation.
+   * triangulations into the output triangulation. If @p copy_manifold_ids is
+   * activated, manifold ids will be copied. Currently, boundary_ids are never
```
activated -> set to @p true

Also, remove `currently` -- we should be documenting the current state of the library, not make reference to past or future behavior

bangerth commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -984,11 +984,11 @@ namespace GridGenerator
+   * triangulations into the output triangulation. If @p copy_manifold_ids is
+   * activated, manifold ids will be copied. Currently, boundary_ids are never
+   * copied. In other words, if the two coarse meshes have anything but the
+   * default boundary indicators, then you will currently have to set boundary
```
same here

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -3809,36 +3812,121 @@ namespace GridGenerator
         this_cell.material_id = cell->material_id();
-        cells.push_back(this_cell);
+        this_cell.manifold_id = cell->manifold_id();
+        cells.push_back(std::move(this_cell));
```
use `emplace_back`. I *think* (but am not sure) that you can then also omit the `std::move` call.

masterleinad replied:
No, `std::move` is necessary to call the move constructor and it doesn't matter whether `push_back` or `emplace_back` is used. (https://stackoverflow.com/questions/26860749/efficiency-of-c11-push-back-with-stdmove-versus-emplace-back-for-already). Also compare https://godbolt.org/g/ebNwx8.

bangerth replied:
great, thank you for checking!

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -3809,36 +3812,121 @@ namespace GridGenerator
         this_cell.material_id = cell->material_id();
-        cells.push_back(this_cell);
+        this_cell.manifold_id = cell->manifold_id();
+        cells.push_back(std::move(this_cell));
```
same here

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -3809,36 +3812,121 @@ namespace GridGenerator
+    SubCellData subcell_data;
+    if (copy_manifold_ids)
+      {
+        if (dim == 2)
```
make this a switch on the dimension.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -3809,36 +3812,121 @@ namespace GridGenerator
+                        line->vertex_index(1) + offset;
+                      boundary_line.manifold_id = line->manifold_id();
+                      subcell_data.boundary_lines.push_back(
+                        std::move(boundary_line));
```
emplace_back

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -3809,36 +3812,121 @@ namespace GridGenerator
+
+                      boundary_quad.manifold_id = face->manifold_id();
+                      subcell_data.boundary_quads.push_back(
+                        std::move(boundary_quad));
```
and here

bangerth commented on source/grid/grid_tools.cc:
```diff
@@ -586,16 +586,19 @@ namespace GridTools
+      for (auto &vertex_index : cell.vertices)
+        vertex_index = new_vertex_numbers[vertex_index];
+    for (auto &quad : subcelldata.boundary_quads)
+      for (auto &vertex_index : quad.vertices)
```
This is a bug fix in that we simply never did this before, right? Can you write a separate changelog entry for that as well?

bangerth commented on source/grid/tria.cc:
```diff
@@ -2095,27 +2095,34 @@ namespace internal
             if (spacedim == 2)
-              AssertThrow((n_adj_cells >= 1) && (n_adj_cells <= 2), ExcInternalError()) else AssertThrow(
-                (n_adj_cells >= 1) && (n_adj_cells <= 2),
-                ExcMessage("You have a line in your triangulation "
```
Ugh :-(

drwells replied:
It looks like we forgot a semicolon after `AssertThrow` before. That really confuses clang-format.

bangerth replied:
It was on purpose. `Assert` expands into `if (cond) {...}`, so if you want to say
```
  if (something)
    Assert (..., ...)
  else
    ...
```
you need to omit the semicolon or the `else` will no longer be syntactically correct.

bangerth replied:
It might be interesting to see with a script if there are other places/how many other places there are where the indentation script now has things broken up in this awkward way.

drwells replied:
We usually put in the semicolon, though; fortunately none of our compilers are that pedantic.

drwells replied:
If I have some time later I'll try to put together a little script to find these. It would be good to add missing semicolons.

bangerth replied:
What I mean to say is that you *can't* put a semicolon after the `Assert` if there is an `else` following, because then it would expand to
```
 if (something)
  {
    if (cond)
      trigger assertion
  };                 // note semicolon
  else
    ...something else...
```
Here, the `{...}` encloses a block, the semicolon is superfluous and ends an otherwise empty statement, and the `else` is now attached to that empty statement and not the `if` code block any more. That's not legal. So you *can't* have the semicolon in this situation.

bangerth replied:
The obvious workaround, used by @masterleinad, is to enclose the `Assert` in curly braces. This way you can add the semicolon at will.

bangerth commented on source/grid/tria.cc:
```diff
@@ -3099,28 +3106,25 @@ namespace internal
+              AssertThrow(line->manifold_id() == boundary_line->manifold_id,
+                          ExcMessage(
+                            "Duplicate boundary lines are only allowed "
+                            "if they carry the same manifold indicator."));
```
Is it clear at this point that the line is really at the boundary? It seems to me like you're already outside the `if (line->at_boundary())` block.

masterleinad replied:
No, the line is not necessarily on the boundary. Otherwise, I can't copy all the manifold ids.

bangerth replied:
But then the error message and the comment above is wrong :-)

bangerth commented on source/grid/tria.cc:
```diff
@@ -3255,26 +3259,25 @@ namespace internal
                           ExcMessage(
                             "Duplicate boundary quads are only allowed "
-                            "if they carry the same boundary indicator."));
+                            "if they carry the same manifold indicator."));
```
same issue here

bangerth commented:
What happens if the two triangulations use the same manifold ids, but with different manifold objects? That should be easy to make happen if you create the two triangulations using the `GridGenerator` functions which all count up their manifold ids starting at zero.

I don't think there's anything in particular you can do about this because you can't compare manifold objects. But we may at least want to document this situation.

davydden commented:
> What happens if the two triangulations use the same manifold ids

btw, that's the reason I put manifold id as parameters in https://github.com/dealii/dealii/pull/6892 . We should probably update GridGenerator functions with manifolds to have extra argument(s).

masterleinad commented:
> What happens if the two triangulations use the same manifold ids, but with different manifold objects?

Nothing happens. In this PR, `Manifold` objects are not copied.

bangerth commented:
Ah, that makes sense!

masterleinad commented:
/run-tests

drwells commented:
@bangerth Are you satisfied with the present state of the PR? If so lets merge.
